src/service-worker/core.js

```diff
@@ -0,0 +1,306 @@
+    const files = transaction.objectStore('js-files');
+
+    // First check if we have this exact file-version.
+    return files.index('fileVersions').get(id).then(file => {
+      // We have it cached!
+      if (file) {
+        return file.requestVersion;
+      }
+
+      // Do we have any versions of this file cached?
+      // To get the first cached file, we must open a cursor. We won't actually
+      // iterate past the first match, though, since we do not advance the
+      // cursor.
+      return files.index('files').openCursor(requestFile).iterate(cursor => {
+        return cursor.value.version;
+      }).then(versions => {
+        // We have a version in cache (cause we iterated over it)!
+        if (versions.length) {
+          return versions[0];
+        }
+
+        // This is the first request we've seen for this file.
+        return '';
+      });
+    });
+  });
+}
+
+self.addEventListener('install', install => {
+  install.waitUntil(Promise.all([cachePromise, dbPromise]));
+});
+
+// Setup the Fetch listener
+// My assumptions:
+//   - Doc requests one uniform AMP release version for all files, anything
+//     else is malarkey.
+self.addEventListener('fetch', event => {
+  const request = event.request;
+  const clientId = event.clientId;
+  const url = request.url;
+  let response;
+
+  // We only cache CDN JS files, and we need a clientId to do our magic.
+  if (clientId && isCdnJsFile(url)) {
+    const requestFile = basename(url);
+    const requestVersion = ampVersion(url);
+
+    // What version do we have for this client?
+    response = Promise.resolve(clientsMap[clientId]).then(version => {
+      // If we already registered this client, we must always use the same
+      // version.
+      if (version) {
+        return version;
+      }
+
+      // If not, do we have this version cached?
+      return getCachedVersion(requestFile, requestVersion).then(version => {
+        // We have a cached version! Serve it up!
+        if (version) {
+          return version;
+        }
+
+        // Tears! We have nothing cached, so we'll have to make a request.
+        return requestVersion;
+      });
+    }).then(version => {
+      const versionedRequest = version === requestVersion ?
+          request :
+          new Request(versionedUrl(url, version), request);
+      clientsMap[clientId] = version;
+
+      return cache.match(versionedRequest).then(response => {
+        // Cache hit!
+        if (response) {
+          // Now, was it because we served an old cached version or because
+          // they requested this exact version; If we served an old version,
+          // let's get the new one.
+          if (versionedRequest !== request) {
+            fetchAndCache(request);
+          }
+
+          return response;
+        }
+
+        // If not, let's fetch and cache the request.
+        return fetchAndCache(versionedRequest);
+      });
+    });
+  } else {
+    response = fetch(request);
```

Comment:
Maybe short circuit above instead of the `else` with `event.respondWith(fetch(request))`